pkg/dataobj/uploader/uploader.go

```diff
@@ -0,0 +1,93 @@
+package uploader
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"flag"
+	"fmt"
+	"time"
+
+	"github.com/grafana/dskit/backoff"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/thanos-io/objstore"
+)
+
+type Config struct {
+	// SHAPrefixSize is the size of the SHA prefix used for splitting object storage keys
+	SHAPrefixSize int
+}
+
+// RegisterFlagsWithPrefix registers flags with the given prefix.
+func (cfg *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.IntVar(&cfg.SHAPrefixSize, prefix+"sha-prefix-size", 2, "The size of the SHA prefix to use for generating object storage keys for data objects.")
+}
+
+func (cfg *Config) Validate() error {
+	if cfg.SHAPrefixSize <= 0 {
+		return fmt.Errorf("SHAPrefixSize must be greater than 0")
+	}
+	return nil
+}
+
+type Uploader struct {
+	SHAPrefixSize int
+	bucket        objstore.Bucket
+	tenantID      string
+	metrics       *metrics
+}
+
+func New(cfg Config, bucket objstore.Bucket, tenantID string) *Uploader {
+	metrics := newMetrics(cfg.SHAPrefixSize)
+
+	return &Uploader{
+		SHAPrefixSize: cfg.SHAPrefixSize,
+		bucket:        bucket,
+		tenantID:      tenantID,
+		metrics:       metrics,
+	}
+}
+
+func (d *Uploader) RegisterMetrics(reg prometheus.Registerer) error {
+	return d.metrics.register(reg)
+}
+
+func (d *Uploader) UnregisterMetrics(reg prometheus.Registerer) {
+	d.metrics.unregister(reg)
+}
+
+// getKey determines the key in object storage to upload the object to, based on our path scheme.
+func (d *Uploader) getKey(object *bytes.Buffer) string {
+	sum := sha256.Sum224(object.Bytes())
+	sumStr := hex.EncodeToString(sum[:])
+
+	return fmt.Sprintf("tenant-%s/objects/%s/%s", d.tenantID, sumStr[:d.SHAPrefixSize], sumStr[d.SHAPrefixSize:])
+}
+
+// Upload uploads an object to the configured bucket and returns the key.
+func (d *Uploader) Upload(ctx context.Context, object *bytes.Buffer) (string, error) {
+	timer := prometheus.NewTimer(d.metrics.uploadTime)
+	defer timer.ObserveDuration()
+
+	objectPath := d.getKey(object)
+
+	backoff := backoff.New(ctx, backoff.Config{
+		MinBackoff: 100 * time.Millisecond,
+		MaxBackoff: 10 * time.Second,
+		MaxRetries: 20,
+	})
+
+	var lastErr error
+	for backoff.Ongoing() {
+		err := d.bucket.Upload(ctx, objectPath, bytes.NewReader(object.Bytes()))
+		if err == nil {
+			return objectPath, nil
+		}
+		lastErr = err
+		backoff.Wait()
+	}
+
+	d.metrics.uploadFailures.Inc()
```

Comment:
Do we want to increment this in the retry loop or only at the very end? (Genuine question, I'm not sure what the tradeoffs would be)

Reply:
I'm opting for the loop. We shouldn't get multiple failures in a row realy. We used to backoff forever, in which case we'd never increment a "once per upload" metric, but now I've set an upper limit to avoid getting completely stuck. 

I think it is better to know we are erroring a lot & making calls to object storage which incurs cost. The alternative is hiding the full scale of the costs we are incurring. But I'm on the fence and can see this both ways: Maybe we don't care about occasional failures and only want to record the failures after multiple attempts.

I'm going to leave this as it is until we get more experience running it.